Add github action and adjust all code to hlint/stylish

paweljakubas opened the pull request:
### Issue Number

ADP-2477

### Comments

- [x] Added both `hlint` and `stylish-haskell` GitHub actions, with versions that are compatible with nix build tool versions.
- [x] Adjusted code to comply with `hlint` suggestions and satisfy `stylish-haskell`.

paweljakubas committed:
hlint/style

paweljakubas committed:
adjust coin selection

paweljakubas committed:
adjust other libs

paweljakubas committed:
adjust in wallet's api bench exe and extra

paweljakubas committed:
adjust in wallet's integration

paweljakubas committed:
adjust in wallet's src and test

paweljakubas committed:
decompose stylish and hlint

paweljakubas committed:
more hlint

paweljakubas committed:
stylish-haskell fix

paweljakubas committed:
choose 3.3.1 hlint

improvements for hlint

one more try

try one more time

try one more time 1

try one more time 2

paweljakubas committed:
rewrite stylish-haskell

fix relative dirs

test 1

test 2

use curl -L

post rebase fix

style.yaml fix

Update lib/wallet/bench/latency-bench.hs

Co-authored-by: Jonathan Knowles <jonathan.knowles@iohk.io>

Update lib/wallet/bench/latency-bench.hs

Co-authored-by: Jonathan Knowles <jonathan.knowles@iohk.io>

Update lib/wallet/integration/src/Test/Integration/Framework/DSL.hs

Co-authored-by: Jonathan Knowles <jonathan.knowles@iohk.io>

Update lib/wallet/integration/src/Test/Integration/Scenario/API/Shelley/Settings.hs

Co-authored-by: Jonathan Knowles <jonathan.knowles@iohk.io>

Update lib/wallet/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs

Co-authored-by: Jonathan Knowles <jonathan.knowles@iohk.io>

Update prototypes/light-mode-test/src/Light/ReadBlocks.hs

Co-authored-by: Jonathan Knowles <jonathan.knowles@iohk.io>

Unisay commented on scripts/exe-from-github.sh:
```diff
@@ -19,7 +19,11 @@
 #   PACKAGE              Name of the release package
 #   PLATFORM             Target platform of the executable
 #   VERSION              Executable version number
+<<<<<<< HEAD
```
This doesn't look right

paweljakubas replied:
indeed, rebase artefact. thanks, fixed

jonathanknowles commented on lib/wallet/bench/latency-bench.hs:
```diff
@@ -415,8 +414,8 @@ walletApiBench capture ctx = do
-        arbitraryStake = Just $ ada 10000
-            where ada = Coin . (1000*1000*)
+        arbitraryStake = Just $ ada 10_000
+            where ada = Coin . (1_000*1_000*)
```
```suggestion
            where ada = Coin . (1_000_000 *)
```

jonathanknowles commented on lib/wallet/bench/latency-bench.hs:
```diff
@@ -433,7 +432,7 @@ withShelleyServer
     ctx <- newEmptyMVar
     let setupContext np baseUrl = do
-            let sixtySeconds = 60*1000*1000 -- 60s in microseconds
+            let sixtySeconds = 60*1_000*1_000 -- 60s in microseconds
```
```suggestion
            let sixtySeconds = 60 * 1_000_000 -- 60s in microseconds
```

jonathanknowles commented on lib/wallet/integration/src/Test/Integration/Framework/DSL.hs:
```diff
@@ -2144,7 +2144,7 @@ json = aesonQQ
 arbitraryStake :: Maybe Coin
 arbitraryStake = Just $ ada 10_000_000_000
-  where ada = Coin . (1000*1000*)
+  where ada = Coin . (1_000*1_000*)
```
```suggestion
  where ada = Coin . (1_000_000 *)
```

jonathanknowles commented on lib/wallet/integration/src/Test/Integration/Scenario/API/Shelley/Settings.hs:
```diff
@@ -90,4 +90,4 @@ spec = describe "SHELLEY_SETTINGS" $ do
 arbitraryStake :: Maybe Coin
 arbitraryStake = Just $ ada 10_000
-  where ada = Coin . (1000*1000*)
+  where ada = Coin . (1_000*1_000*)
```
```suggestion
  where ada = Coin . (1_000_000 *)
```

jonathanknowles commented on lib/wallet/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs:
```diff
@@ -1122,11 +1122,10 @@ instance (Arbitrary a, Ord a) => Arbitrary (NonSingletonRange a) where
-    shrink (NonSingletonRange r) = catMaybes $
-        makeNonSingletonRangeValid . NonSingletonRange <$> shrink r
+        head . mapMaybe (makeNonSingletonRangeValid . NonSingletonRange)
+               <$> infiniteList
```
```suggestion
        head . mapMaybe (makeNonSingletonRangeValid . NonSingletonRange)
            <$> infiniteList
```

jonathanknowles commented on .github/workflows/style.yml:
```diff
@@ -0,0 +1,50 @@
+        os: [ubuntu-20.04]
+    runs-on: ${{ matrix.os }}
+    steps:
+    - name: 📥 Checkout repository
```
```suggestion
    - name: 📥 Check out repository
```

paweljakubas replied:
done

jonathanknowles commented on prototypes/light-mode-test/src/Light/ReadBlocks.hs:
```diff
@@ -137,15 +136,13 @@ lightSync follower = do
+                 old <- rollBackward follower $ fromBlock target
+                 nextBlockFrom old)
+         else
+             fromBlock <$> BF.getLatestBlock)
```
I think the multi-line parenthesis is much harder to read than the original block layout.

Can we keep the old pattern match here?
```suggestion
        case b of
            False -> fromBlock <$> BF.getLatestBlock
            True -> do
                -- fall back to behind the stability window
                target <- BF.getBlock . Left $ height - stabilityWindow
                old <- rollBackward follower $ fromBlock target
                nextBlockFrom old
```

paweljakubas replied:
sounds reasonable

jonathanknowles reviewed (changes requested):
Hi @paweljakubas 

Thanks for adding this GitHub action!

I do wonder about some of the changes though. Would it not have been possible to add some "surgical" `HLINT` statements to tops of files to avoid having to create such a large PR?

I like the use of `NumericUnderscores` though, as you know!

Two things I would like to see:
1. Could you link the PR to a JIRA ticket? (So we can track the work we're doing to rebuild our CI system.)
2. Could you fill out the PR description, to briefly state what you've done, and why?
3. Are you able to arrange these commits by **_type_** of change? For example:
    - Commit 1: `Use NumericUnderscores.`
    - Commit 2: `Use void instead of "() <$".`
    - Commit 3: `Apply eta reductions.`
 
 I would say that it's much easier to review if we can see a single commit per type of change, rather than separating commits arbitrarily by area of the code, as there's less context switching for reviewers (and people reading the commit history).
 
Alternatively, if separating commits in this way is too much work, then I would propose squashing the commits that make changes to the code, as there's not much benefit to separating them out as they are now.

jonathanknowles reviewed (approved):
Hi @paweljakubas 

Many thanks for filling out the PR description and linking to the JIRA ticket (https://input-output.atlassian.net/browse/ADP-2477), I appreciate it! 🙇🏻 

I took the liberty of assigning https://input-output.atlassian.net/browse/ADP-2477 to the [current sprint](https://input-output.atlassian.net/jira/software/c/projects/ADP/boards/231). Hope that's okay with you! My reason for doing this is so that we can track this item of work and account for it together with all the other pieces of work that our team has had to pick up in the wake of our old CI system going away. I think it's important that your work is noticed and recognised!

As for the commit structure, I think it's fine, and understand that it's too painful to reorganise at this stage. Though perhaps squashing some of the commits together might leave a cleaner history?

Thanks again for your hard work in making this PR. 💯

jonathanknowles commented on lib/coin-selection/test/spec/Cardano/CoinSelection/CollateralSpec.hs:
```diff
@@ -937,11 +937,11 @@ unitTests_numberOfSubsequencesOfSize_withinBounds = unitTests
+        , (100,  2,                     4_950)
+        , (100,  4,                 3_921_225)
+        , (100,  8,           186_087_894_300)
+        , (100, 16, 1_345_860_629_046_814_650)
```
👍🏻

Anviking reviewed (approved):
🎉 

Just, with the .hs changes, could we wait for #3627 and see buildkite green for added re-assurance? I _think_ the docker check is compiling the code, at least to some extent.

paolino reviewed (approved):
This is very nice! Thanks for syntax-shaking the tree! ... Fourmolu ... I can hear you coming!

paweljakubas committed:
add HLint ignore

paweljakubas committed:
tmp check

paolino committed:
fix style scripts

add package requirements
use paolino hosted runner

fix style scripts

use paolino hosted runner

paolino committed:
fix exe-from-github.sh

piotr-iohk commented on .github/workflows/style.yml:
```diff
@@ -0,0 +1,49 @@
+
+jobs:
+  build:
+    runs-on: self-hosted
```
why self-hosted?

piotr-iohk commented on .github/workflows/style.yml:
```diff
@@ -0,0 +1,49 @@
+        RELEASE: v3.3.1
+        VERSION: 3.3.1
+      run: |
+        apt install -y libtinfo5 
```
seem redundant since there is the same command on "📥 Add packages"?

paolino committed:
fix style scripts

add package requirements
use paolino hosted runner

fix style scripts

use paolino hosted runner

paweljakubas committed:
revert script
